discord-scripts/thread-management.ts

```diff
@@ -0,0 +1,134 @@
+import { Channel, ChannelType, Client } from "discord.js"
+
+// Emoji used to suggest a thread.
+const THREAD_EMOJI = "🧵"
+// Category that is treated as recreational, i.e. the rules don't apply baby.
+const RECREATIONAL_CATEGORY_ID = "1079492118692757605"
+
+function isInRecreationalCategory(
+  channel: Channel | undefined | null,
+): boolean {
+  if (
+    channel === undefined ||
+    channel === null ||
+    channel.isDMBased() ||
+    channel.parent === null
+  ) {
+    return false
+  }
+
+  // Channel is inside a category.
+  if (channel.parent.type === ChannelType.GuildCategory) {
+    return channel.parent.id === RECREATIONAL_CATEGORY_ID
+  }
+
+  // Channel's parent is inside a category; this applies to thread channels.
+  if (channel.parent.parent !== null) {
+    return channel.parent.id === RECREATIONAL_CATEGORY_ID
+  }
+
+  return false
+}
+
+export default function manageThreads(discordClient: Client) {
+  // When a thread is created, join it.
+  //
+  // Additionally, quietly tag a role so that all members of it are subscribed
+  // to the thread (they may later leave the thread to opt out). The role that
+  // is tagged is, in order:
+  //
+  // - If the containing channel's category is recreational, no role.
+  // - If the containnig channel has a role with a matching name, that role
+  //   (e.g., a message to #tech will tag a Tech role if it exists).
+  // - If the containing channel's category has a role with a matching name, that role
+  //   (e.g., a message to #taho-standup inside the Taho category will tag the
+  //   Taho role if it exists).
+  // - If the containing channel's category is General and the channel is
+  //   #main, @everyone.
+  discordClient.on("threadCreate", async (thread) => {
+    await thread.join()
+
+    if (isInRecreationalCategory(thread)) {
+      return
+    }
+
+    const { guild: server, parent: containingChannel } = thread
+
+    const placeholder = await thread.send("<placeholder>")
+
+    const matchingRole = server.roles.cache.find(
+      (role) =>
+        role.name.toLowerCase() === containingChannel?.name.toLowerCase(),
+    )
+
+    if (matchingRole !== undefined) {
+      await placeholder.edit(matchingRole.toString())
+      return
+    }
+
+    const categoryChannel = containingChannel?.parent
+    const categoryMatchingRole = server.roles.cache.find(
+      (role) => role.name.toLowerCase() === categoryChannel?.name.toLowerCase(),
+    )
+
+    if (categoryMatchingRole !== undefined) {
+      await placeholder.edit(categoryMatchingRole.toString())
+      return
+    }
+
+    if (
+      categoryChannel?.name?.toLowerCase()?.endsWith("general") === true &&
+      containingChannel?.name?.toLowerCase()?.endsWith("main") === true
+    ) {
+      await placeholder.edit(server.roles.everyone.toString())
```

Comment:
is this @-everyone? if so, i don't think the @-everyone trick works; have to give everyone the same role and use that instead

Reply:
Does "not work" mean people don't get joined correctly? Might explain the issue @michalinacienciala noticed @ https://discord.com/channels/1079490158639976609/1090503843508408330/1090642962729414656 .

Reply:
yeah - as far as I know, editing into @-everyone doesn't add folks

Reply:
@-everyone:
<img width="1146" alt="image" src="https://user-images.githubusercontent.com/1045160/228591160-958ac8a7-792a-4972-8c26-cf3ade7308b1.png">

@-keep:
<img width="1133" alt="image" src="https://user-images.githubusercontent.com/1045160/228591301-5e8845cd-0578-40a2-8b49-f59761d70368.png">